drud/ddev
# Improved management of settings files, introduce settings.ddev.php

## The Problem/Issue/Bug:
#468, improved management of settings files.

## How this PR Solves The Problem:
This PR creates settings.ddev.php, placing ddev-specific configurations in this file to preserve user settings.  The settings.php file is only modified to include settings.ddev.php.

## Manual Testing Instructions:
[Randy's comment here](https://github.com/drud/ddev/issues/468#issuecomment-395548090) describes the desired behavior and was used as the template for the changes in this PR. 

For Drupal 6/7/8 and Backdrop, begin with the situation in which no settings files exist.
- Begin with no settings.php file and no settings.ddev.php file, and no .gitignore in the directory containing the settings files,
- `ddev config`
- Ensure a settings.php has been created that does nothing but include settings.ddev.php
- Ensure a settings.ddev.php has been created that contains ddev-generated configurations
- Ensure a .gitignore has been created that ignores settings.ddev.php

Now, test the situation in which a settings.php file already exists but does not include settings.ddev.php
- Begin with no settings.ddev.php and a settings.php file that has contents, but does not include settings.ddev.php
- `ddev config`
- Ensure settings.php has been modified to include settings.ddev.php
- Ensure settings.ddev.php has been created
- Ensure a .gitignore has been created that ignores settings.ddev.php

Finally, test the situation in which settings.pho and .gitignore already exist and won't be modified, and a settings.ddev.php will be overwritten.
- Begin with a settings.php that includes settings.ddev.php, an existing settings.ddev.php filled with "junk" data, and an existing user-managed .gitignore in the settings directory
- `ddev config`
- Ensure settings.php has not been modified to include settings.ddev.php
- Ensure settings.ddev.php has been overwritten and contains the settings as generated by ddev
- Ensure .gitginore has not been overwritten

## Automated Testing Overview:
Several tests have been added to ensure the desired creation/modification behavior of settings.php, settings.ddev.php, and .gitignore in settings_test.go.

## Related Issue Link(s):

## Release/Deployment notes:
